xml_converter/src/xml_converter.cpp

```diff
@@ -414,45 +408,46 @@ bool filename_comp(string a, string b) {
     return lowercase(a) < lowercase(b);
 }
 
-vector<string> get_xml_files(string directory) {
+vector<string> get_xml_files(string directory, string data_directory) {
     vector<string> files;
-    vector<string> subfolders;
-    string path;
     DIR* dir = opendir(directory.c_str());
-    struct dirent* entry = readdir(dir);
+    struct dirent* entry;
     while ((entry = readdir(dir)) != NULL) {
-        if (entry->d_type == DT_DIR && has_suffix(directory, ".") == false) {
-            path = directory + "/";
-            path += entry->d_name;
-            subfolders.push_back(path);
-        }
-        else if (has_suffix(entry->d_name, ".xml")) {
-            path = directory + "/";
-            path += entry->d_name;
-            files.push_back(path);
-        }
-        else {
-            continue;
+        string filename = entry->d_name;
+        if (filename != "." && filename != "..") {
+            string path = directory + "/" + filename;
+            if (entry->d_type == DT_DIR) {
+                string new_folder = data_directory + "/" + filename;
+                if (mkdir(new_folder.c_str(), 0700) == -1 && errno != EEXIST) {
+                    cout << "Error making " << new_folder << endl;
+                    continue;
+                }
+                vector<string> subfiles = get_xml_files(path, new_folder);
+                for (string subfile : subfiles) {
+                    cout << subfile << endl;
+                    files.push_back(subfile);
+                }
+            }
+            else if (entry->d_type == DT_REG) {
+                if (has_suffix(filename, ".xml")) {
+                    files.push_back(path);
+                }
+                else if (has_suffix(filename, ".trl") == false) {
+                    string new_path = data_directory + "/" + filename;
```

Comment:
You dont need to read the files and write them, you can use the OS copy file functionality instead.